Add runcron admin management command

sounak98 opened the pull request:
This PR will add the `runcron` admin management command in django to run scheduled tasks.

Fixes #31.

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Run command `python manage.py runcron <command> --num_workers=5 --timeout=10`

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

sounak98 committed:
Add basic structure of runcron admin command

sounak98 commented:
@botanicvelious check the structure of the management command, does it look good? I will now proceed with creating the `Scheduler` model.

botanicvelious commented:
general idea looks good, the 2 commands should be "build_items" and "process_items" however. Also the -t and -n only would be arguments for "process_items" so that would also be good to change or explicitly ignore. Lastly should set it as a default to run build then process if neither one is specified.

sounak98 commented:
@botanicvelious have updated the structure to allow the specified commands `build_items` and `process_items` and take the former as the default one when nothing is mentioned.

sounak98 committed:
Add Scheduler model

botanicvelious reviewed (changes requested):
It would be a good idea to use a TextField rather than a char for the data column, as 1000 chars could likely not be enough for some items so just use a TextField so you never have to worry about it.

Also, it should build, then run not just build if there are no additional arguments.

Lastly please dont commit changes to the .db files, that way we dont run into conflicts. You can set this by .git/info/exclude and git update-index --skip-worktree [<file>...]

sounak98 committed:
Extend runcron management command

sounak98 commented:
- [x] `data` field is updated to `TextField` from `CharField` (was already going to do this! 😅)
- [x] the `db` files have been removed from the commits
- [x] the management command `runcron` supports two subcommands, viz. `process_items` and `build_items`. The `process_items` takes up the unattended scheduled tasks and runs them on a thread pool before marking them successful.

The subcommands of the `runcron` management command are:
- `python manage.py runcron process_items --num_workers=5 --timeout=10`
- `python manage.py runcron build_items`

@botanicvelious I have left the util functions `send_email` and `send_irc_msg` empty, and also the `build_items` function; all of which can be worked on in different PRs.

botanicvelious commented:
Overall looks good, you missed that 

```
manage.py runcron
Build Items
```

Should build then run, not just build. So the output would be 

```
manage.py runcron
Build Items
No scheduled tasks
```

Also, the code looks to not use the timeout on a worker process at all. You could easily test your process timeout function by waiting an amount of time greater than the timeout in for example send_email and waiting a bit less in the send_irc_msg.

Lastly, the success doesn't seem to take into account if the processed item was actually successful, just looks to assume it was. Thinking about it would also be useful to have a "failure_message" field for when there is a failure so that can be logged or notify someone etc. Additionally, it looks like the default is failed, when the default should be null, then updated to success or failure based on if the process completes successfully.

sounak98 committed:
Add timeout and error management in runcron tasks

sounak98 commented:
@botanicvelious `scheduler.success == True` means that the scheduled command has been successfully executed. In all other scenarios `scheduler.success == False`. I have added a log for the errors caused while executing the commands (if any). If the execution fails `scheduler.success` remains `False`.

My bad, I forgot about the timeout. I've added it.

botanicvelious commented:
The issue with having is starting false and stay false is that you then have no way of knowing if it ran and failed as it says false in either case. That is why it's typically best to have a null that is changed to the proper value. This way you can run the "process_items" against anything where status = null, then you know if something failed it is false if it succeeded its true, and you can easily have a process in "build_items" to make it so that it re-adds anything that is failed lets say up to 3 times or whatever. This way you have a full log of everything without having to rely on additional logging systems or otherwise.

If you want to get on irc I am online currently and we can chat on the best way to handle this.

sounak98 committed:
Add error logging in Scheduler model

sounak98 commented:
I agree with you @botanicvelious, I've updated the `Scheduler` to have a field for the last error and also `scheduler.success` is set to `None` when the `Scheduler` object is being created.

botanicvelious reviewed (commented):
> Best practice in Django is to have a BooleanField accept only the two values: true and false. If you need to accept NULL values, you have to use a NullBooleanField.

Actually, my bad I wasn't thinking a default value. Everything looks good.

sounak98 commented:
@botanicvelious django documentation says
> ### NullBooleanField
> Like BooleanField with null=True. Use that instead of this field as it’s likely to be deprecated in a future version of Django.

Oh, I just saw your comment.

botanicvelious commented:
> 
> 
> @botanicvelious django documentation says
> 
> > ### NullBooleanField
> > Like BooleanField with null=True. Use that instead of this field as it’s likely to be deprecated in a future version of Django.

Yes, I realized after that I was not thinking.

botanicvelious commented on gsoc/management/commands/runcron.py:
```diff
@@ -1,10 +1,82 @@
+            scheduler.save()
+
+    def process_items(self, options):
+            schedulers = Scheduler.objects.exclude(success=True)
```
This excludes true, but it needs to actually only include null

sounak98 replied:
So, the processes that have been run before and have failed for some reason doesn't need to be run? Or should we add another flag for that? Something like `--exclude-failed`?

sounak98 replied:
And what should be the default behaviour?

botanicvelious replied:
I would just never run failed stuff, we can always re-add anything that failed as part of the build_items

botanicvelious commented on gsoc/management/commands/runcron.py:
```diff
@@ -1,10 +1,82 @@
+        if options['task']:
+            getattr(self, options['task'])(options)
+        else:
+            self.build_items(options)
```
Shouldnt this be

```Python
self.build_items(options)
self.process_items(options)
```

sounak98 replied:
`self.build_items(options)` run `self.process_items(options)` inside it.

botanicvelious replied:
That would then make it impossible to just build the items and not process them. That could be problematic down the line.

sounak98 replied:
So then on running `runcron build_items` should it process it too?

I guess it shouldn't. Had a misconception.

botanicvelious replied:
>runcron build_items 

would only build

>runcron 

would build and run

sounak98 replied:
The latest updates change the behaviour to exactly this.

botanicvelious commented on gsoc/models.py:
```diff
@@ -3,6 +3,23 @@
+    id = models.AutoField(primary_key=True)
+    command = models.CharField(name='command', max_length=20, choices=commands)
+    data = models.TextField(name='data')
+    success = models.BooleanField(name='success', null=True, default=None)
```
The default value of BooleanField is None when Field.default isn’t defined so is the default really necessary?

botanicvelious commented on gsoc/models.py:
```diff
@@ -3,6 +3,23 @@
+    command = models.CharField(name='command', max_length=20, choices=commands)
+    data = models.TextField(name='data')
+    success = models.BooleanField(name='success', null=True, default=None)
+    last_error = models.CharField(name='last_error', max_length=100, null=True, default=None)
```
A max_length=100 char field for errors may be quite small if there is like a stack trace or something especially. TextField probably the better choice.

sounak98 replied:
Yeah, even I was thinking about this. I'll change this.

botanicvelious reviewed (commented):
Other then these items looks good.

sounak98 committed:
Change error field of Scheduler to TextField

sounak98 committed:
Minor changes

- Change filter argument for Schedulers in runcron process_item task
- Change success field of Scheduler
- Change default runcron action to first build then process

sounak98 commented:
@botanicvelious I have resolved all the reviews. Can you check?

botanicvelious commented:
@sounak98 looks good except for the one 

>`self.build_items(options)` run `self.process_items(options)` inside it.

if you can make that one last change i will merge the code

sounak98 commented:
> if you can make that one last change i will merge the code

@botanicvelious Done!

botanicvelious commented:
@sounak98 gsoc/management/commands/runcron.py is still unresolved?

sounak98 commented:
It's resolved @botanicvelious

botanicvelious commented:
Merged!
